src/index.js

```diff
@@ -1,76 +1,116 @@
+    let recompileInterval;
+    const options = compiler.options;
+    const distPath =
+      get(options, "devServer.static.directory") ||
+      get(options, "output.path") ||
+      "dist";
+
+    const typescriptFolderName = "@mf-typescript";
+    const typesIndexFileName = "__types_index.json";
+
+    const distDir = path.join(distPath, typescriptFolderName);
+    const outFile = path.join(distDir, typesIndexFileName);
+
+    const tsCompilerOptions = {
+      declaration: true,
+      emitDeclarationOnly: true,
+      outDir: path.join(distDir, "/"),
+    };
 
     const federationOptions = options.plugins.find((plugin) => {
-      return plugin.constructor.name === 'ModuleFederationPlugin'
+      return plugin.constructor.name === "ModuleFederationPlugin";
     });
-    const inheritedPluginOptions = get(federationOptions, '_options') || null
+    const inheritedPluginOptions = get(federationOptions, "_options") || null;
 
+    const exposedComponents =
+      this.options?.exposes || inheritedPluginOptions.exposes;
+    const remoteComponents =
+      this.options?.remotes || inheritedPluginOptions.remotes;
 
     const run = () => {
-      const exposedComponents = this.options?.exposes || inheritedPluginOptions.exposes
-      const remoteComponents = this.options?.remotes || inheritedPluginOptions.remotes
-
       if (exposedComponents) {
+        const normalizedFileNames = [];
+
         const fileNames = Object.values(exposedComponents);
-        const typeFiles = fileNames.map(f => {
-          const split = f.split('/')
-          return split[split.length - 1].split('.')[0] + '.d.ts'
-        })
-
-        fs.writeFile(path.join(distPath, '/@mf-typescript/__types_index.json'), JSON.stringify(typeFiles), (e) => {
-          if (e) {
-            console.log('Error saving the types index', e)
+
+        fileNames.forEach((componentFilePath) => {
+          const ext = path.extname(componentFilePath);
+          if (["ts", ".tsx"].includes(ext)) {
+            const normalizedPath = path.resolve(__dirname, componentFilePath);
+
+            normalizedFileNames.push(normalizedPath);
+          } else {
+            throw new Error(
+              `Can not determine file extension, please include file extension for the file ${componentFilePath}`
+            );
           }
-        })
-        const program = ts.createProgram(fileNames, {
-          declaration: true,
-          emitDeclarationOnly: true,
-          outDir: path.join(distPath, "./@mf-typescript/"), // replace this with build directory
         });
 
-        program.emit();
+        const host = ts.createCompilerHost(tsCompilerOptions);
+        const program = ts.createProgram(
+          normalizedFileNames,
+          tsCompilerOptions,
+          host
+        );
+
+        const typeFiles = fileNames.map((f) => {
+          const split = f.split("/");
+          return split[split.length - 1].split(".")[0] + ".d.ts";
```

Comment:
What happens if theres multiple dots like thing.component.tsx. Path.basename or path.parse may be more reliable

Reply:
We cannot actually use this map() as the ts compiler emits the filenames as absolute paths.

addressed this with a different approach.